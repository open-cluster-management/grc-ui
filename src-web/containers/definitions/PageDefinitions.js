/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */
/* eslint-disable react/display-name */

'use strict'
import React from 'react'
import { AcmButton, AcmSecondaryNavItem } from '@stolostron/ui-components'

import { ALL_POLICIES, SINGLE_POLICY, POLICY_STATUS, POLICY_STATUS_HISTORY, POLICY_TEMPLATE_DETAILS } from '../../utils/client/queries'
import config from '../../../server/lib/shared/config'
// eslint-disable-next-line import/no-named-as-default
import GrcView from '../../components/modules/GrcView'
import { PolicyDetailsOverview } from '../../components/modules/PolicyDetailsOverview'
import PolicyStatusView from '../../components/modules/PolicyStatusView'
import PolicyTemplateDetailsView from '../../components/modules/PolicyTemplateDetailsView'
import PolicyStatusHistoryView from '../../components/modules/PolicyStatusHistoryView'
import msgs from '../../nls/platform.properties'
import { checkCreatePermission, checkEditPermission } from '../../utils/CheckUserPermission'

const policiesMsg = 'routes.policies'
const historyMsg = 'table.header.history'

export const getPageDefinition = (props) => {
  const { type } = props
  switch(type) {
    case 'ALL_POLICIES':
      return policiesPage(props)
    case 'SINGLE_POLICY':
      return policyDetailsPage(props)
    case 'POLICY_STATUS':
      return policyStatusPage(props)
    case 'POLICY_TEMPLATE_DETAILS':
      return policyTemplateDetailsPage(props)
    case 'POLICY_STATUS_HISTORY':
      return policyStatusHistoryPage(props)
  }
  return null
}

const createBtn = ({ userAccess, history, locale }) => {
  return (
    <AcmButton key='create-policy' id='create-policy' isDisabled={checkCreatePermission(userAccess)===0}
      tooltip={msgs.get('error.permission.disabled', locale)}
      onClick={() => history.push(`${config.contextPath}/create`)}>
      {msgs.get('routes.create.policy', locale)}
    </AcmButton>
  )
}

const editBtn = ({ userAccess, history, locale, name, namespace }) => {
  return (
    <AcmButton key='edit-policy' id='edit-policy' isDisabled={checkEditPermission(userAccess)===0}
      tooltip={msgs.get('error.permission.disabled', locale)}
      onClick={() => history.push(`${config.contextPath}/all/${namespace}/${name}/edit`)}>
      {msgs.get('routes.edit.policy', locale)}
    </AcmButton>
  )
}

const detailsNav = ({ history, locale, name, namespace }) => {
  const url = `${config.contextPath}/all/${namespace}/${name}`
  return (
    <AcmSecondaryNavItem key='details' isActive={history.location.pathname===url}
      onClick={() => history.push(url)}>
      {msgs.get('tabs.details', locale)}
    </AcmSecondaryNavItem>
  )
}

const statusNav = ({ history, locale, name, namespace }) => {
  const url = `${config.contextPath}/all/${namespace}/${name}/status`
  return (
    <AcmSecondaryNavItem key='status' isActive={history.location.pathname===url}
      onClick={() => history.push(url)}>
      {msgs.get('tabs.status', locale)}
    </AcmSecondaryNavItem>
  )
}

const policiesPage = ({ locale }) => {
  return {
    id: 'policies',
    title: msgs.get('routes.grc', locale),
    query: ALL_POLICIES,
    refreshControls: true,
    buttons: [ createBtn ],
    children: (props) => <GrcView {...props} />
  }
}

const policyDetailsPage = ({ name, namespace, locale }) => {
  return {
    id: 'policy-details',
    title: name,
    query: SINGLE_POLICY,
    query_variables: { name, namespace },
    refreshControls: true,
    breadcrumb: [
      { text: msgs.get(policiesMsg, locale), to: config.contextPath },
      { text: name, to: name }
    ],
    navigation: [
      detailsNav,
      statusNav
    ],
    buttons: [ editBtn ],
    children: (props) => <PolicyDetailsOverview {...props} />
  }
}

const policyStatusPage = ({ name, namespace, locale }) => {
  return {
    id: 'policy-status',
    title: name,
    query: POLICY_STATUS,
    query_variables: { policyName: name, hubNamespace: namespace },
    refreshControls: true,
    breadcrumb: [
      { text: msgs.get(policiesMsg, locale), to: config.contextPath },
      { text: name, to: name }
    ],
    navigation: [
      detailsNav,
      statusNav
    ],
    buttons: [ editBtn ],
    children: (props) => <PolicyStatusView {...props} />
  }
}

const policyTemplateDetailsPage = ({ name, namespace, cluster, apiGroup, version, kind, template, locale }) => {
  const selfLink = `/apis/${apiGroup}/${version}/namespaces/${cluster}/${kind}/${template}`
  return {
    id: 'policy-template-details',
    title: template,
    query: POLICY_TEMPLATE_DETAILS,
    query_variables: { name:template, cluster, kind, selfLink },
    refreshControls: true,
    breadcrumb: [
      { text: msgs.get(policiesMsg, locale), to: config.contextPath },
      { text: name, to: `${config.contextPath}/all/${namespace}/${name}`},
      { text: msgs.get('table.header.status', locale), to: `${config.contextPath}/all/${namespace}/${name}/status` },
      { text: template, to: template }
    ],
    children: (props) => <PolicyTemplateDetailsView {...props} selfLink={selfLink} />
  }
}

const policyStatusHistoryPage = ({ name, namespace, cluster, template, locale }) => {
  return {
    id: 'policy-status-history',
    title: msgs.get(historyMsg, locale),
    query: POLICY_STATUS_HISTORY,
    query_variables: { policyName: name, hubNamespace: namespace, cluster, template },
    refreshControls: true,
    breadcrumb: [
      { text: msgs.get(policiesMsg, locale), to: config.contextPath },
      { text: name, to: `${config.contextPath}/all/${namespace}/${name}`},
      { text: msgs.get('table.header.status', locale), to: `${config.contextPath}/all/${namespace}/${name}/status`},
      { text: msgs.get(historyMsg, locale), to: msgs.get(historyMsg, locale) }
    ],
    children: (props) => <PolicyStatusHistoryView {...props} cluster={cluster} template={template} />
  }
}

