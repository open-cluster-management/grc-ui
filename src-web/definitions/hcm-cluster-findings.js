/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'
import _ from 'lodash'

export default {
  defaultSortField: 'cluster',
  primaryKey: 'cluster',
  secondaryKey: 'metadata.namespace',
  tableActions: [
    'table.actions.finding.clusterFindings.sidepanel',
  ],
  tableKeys: [
    {
      msgKey: 'table.header.cluster.name',
      resourceKey: 'cluster',
    },
    {
      msgKey: 'overview.recent.activity.severity.high',
      resourceKey: 'severity',
    },
    {
      msgKey: 'table.header.highSeverity',
      resourceKey: 'highSeverity',
      transformFunction: getHighSeverityObjTostring,
    },
  ],
  clusterFindingSidePanel: {
    headerRows: ['', 'table.header.findingName', 'table.header.severity'],
    subHeaders: ['tabs.detail', ''],
    rows: [
      {
        cells: [
          {
            resourceKey: 'shortDescription',
          },
          {
            resourceKey: 'finding.severity',
          }
        ]
      }
    ]
  },
}

export function getHighSeverityObjTostring(item) {
  const highSeverityMap = _.get(item, 'highSeverity')
  let highSeverityString = ''
  if (highSeverityMap) {
    Object.keys(highSeverityMap).forEach((key) => {
      highSeverityString = `${highSeverityString} ${key} (${highSeverityMap[key]}),`
    })
  }
  highSeverityString = highSeverityString.trim()
  return highSeverityString.substring(0, highSeverityString.length-1)
}
