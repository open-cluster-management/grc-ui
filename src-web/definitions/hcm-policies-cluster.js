/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */

'use strict'

import React from 'react'
import TruncateText from '../components/common/TruncateText'
import config from '../../lib/shared/config'

export default {
  defaultSortField: 'cluster',
  primaryKey: 'cluster',
  secondaryKey: 'metadata.namespace',
  tableActions: [
    'table.actions.policy.clusters.sidepanel',
    'table.actions.launch.cluster'
  ],
  tableKeys: [
    {
      msgKey: 'table.header.cluster.name',
      resourceKey: 'cluster',
      transformFunction: createClusterLink,
    },
    {
      msgKey: 'table.header.cluster.namespace',
      resourceKey: 'namespace',
    },
    {
      msgKey: 'table.header.violation',
      resourceKey: 'violation',
    },
    {
      msgKey: 'table.header.violated',
      resourceKey: 'nonCompliant',
      transformFunction: getClusterViolationLabels,
    }
  ],
  clusterViolatedSidePanel: {
    headerRows: ['', 'table.header.policy.name', 'table.header.rule.violation', 'table.header.control'],
    subHeaders: ['table.header.name', 'table.header.message', 'table.header.reason'],
    rows: [
      {
        cells: [
          {
            resourceKey: 'metadata.name',
          },
          {
            resourceKey: 'violatedNum',
          },
          {
            resourceKey: `metadata.annotations["${config.mcmPolicyPrefix}/controls"]`,
          }
        ]
      }
    ]
  },
}

export function getClusterViolationLabels(item) {
  return getTruncatedText(item.nonCompliant.join(', '))
}

export function getTruncatedText(item){
  return <TruncateText text={item} />
}

export function createClusterLink(item){
  if (item && item.cluster && item.namespace) {
    return <a href={`${config.clusterContextPath}/${item.namespace}/${item.cluster}`}>{item.cluster}</a>
  }
  else if (item && item.cluster) {
    return item.cluster
  }
  return '-'
}
