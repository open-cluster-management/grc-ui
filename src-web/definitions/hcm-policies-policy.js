/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import { Link } from 'react-router-dom'
import config from '../../lib/shared/config'
import {getCategories, getControls, getStandards} from './hcm-compliances'

export default {
  defaultSortField: 'metadata.name',
  primaryKey: 'metadata.name',
  secondaryKey: 'metadata.namespace',
  tableActions: [
    'table.actions.policy.policies.sidepanel',
    'table.actions.edit',
    'table.actions.disable',
    'table.actions.remove',
  ],
  tableKeys: [
    {
      msgKey: 'table.header.policy.name',
      resourceKey: 'metadata.name',
      transformFunction: createComplianceLink,
    },
    {
      msgKey: 'table.header.namespace',
      resourceKey: 'namespace',
    },
    {
      msgKey: 'table.header.remediation',
      information: 'grc.remediation.tooltip',
      resourceKey: 'remediation',
    },
    {
      msgKey: 'table.header.cluster.violation',
      resourceKey: 'clusterCompliant',
    },
    {
      msgKey: 'table.header.standards',
      resourceKey: `metadata.annotations["${config.mcmPolicyPrefix}/standards"]`,
      transformFunction: getStandards,
    },
    {
      msgKey: 'table.header.controls',
      resourceKey: `metadata.annotations["${config.mcmPolicyPrefix}/controls"]`,
      transformFunction: getControls,
    },
    {
      msgKey: 'table.header.categories',
      resourceKey: `metadata.annotations["${config.mcmPolicyPrefix}/categories"]`,
      transformFunction: getCategories
    },
  ],
  policyViolatedSidePanel: {
    // title: 'policy.violated.cluster',
    headerRows: ['', 'table.header.cluster.name', 'table.header.rule.violation', ''],
    subHeaders: ['table.header.name', 'table.header.message', 'table.header.reason'],
    rows: [
      {
        cells: [
          {
            resourceKey: 'metadata.name',
          },
          {
            resourceKey: 'violated',
          },
          {
            resourceKey: 'spec.consoleURL',
          }
        ]
      }
    ]
  },
}

export function createComplianceLink(item = {}, ...param){
  if (param[2]) {
    return item.metadata.name
  } else if (item && item.metadata)
  {
    if (item.raw.kind === 'Compliance') {
      return <Link to={`${config.contextPath}/all/${encodeURIComponent(item.metadata.name)}`}>{item.metadata.name} (Deprecated)</Link>
    }
    else {
      return <Link to={`${config.contextPath}/all/${encodeURIComponent(item.metadata.namespace)}/${encodeURIComponent(item.metadata.name)}`}>{item.metadata.name}</Link>
    }
  }
  return undefined
}

