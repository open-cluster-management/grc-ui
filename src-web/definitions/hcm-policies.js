/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import lodash from 'lodash'
import msgs from '../../nls/platform.properties'
import {getAge, getLabelsToList} from '../../lib/client/resource-helper'
import { Link } from 'react-router-dom'
import resources from '../../lib/shared/resources'
import StatusField from '../components/common/StatusField'
import config from '../../lib/shared/config'

resources(() => {
  require('../../scss/table.scss')
})

export default {
  defaultSortField: 'metadata.name',
  primaryKey: 'metadata.name',
  secondaryKey: 'metadata.namespace',
  policyRules: {
    title: 'table.header.rules',
    defaultSortField: 'ruleUID',
    normalizedKey: 'ruleUID',
    resourceKey: 'rules',
    tableKeys: [
      {
        msgKey: 'table.header.name',
        resourceKey: 'ruleUID',
        key: 'ruleUID',
      },
      {
        msgKey: 'table.header.templateType',
        resourceKey: 'templateType',
        key: 'templateType',
      },
      {
        msgKey: 'table.header.complianceType',
        resourceKey: 'complianceType',
        key: 'complianceType',
      },
      {
        msgKey: 'table.header.apiGroups',
        resourceKey: 'apiGroups',
        key: 'apiGroups',
        transformFunction: getAPIGroups
      },
      {
        msgKey: 'table.header.ruleVerbs',
        resourceKey: 'verbs',
        key: 'verbs',
        transformFunction: getRuleVerbs
      },
      {
        msgKey: 'table.header.resources',
        resourceKey: 'resources',
        key: 'resources',
      },
    ],
  },
  policyRoleTemplates: {
    title: 'table.header.roleTemplates',
    defaultSortField: 'name',
    normalizedKey: 'name',
    resourceKey: 'roleTemplates',
    tableKeys: [
      {
        msgKey: 'table.header.name',
        resourceKey: 'name',
        key: 'name',
      },
      {
        msgKey: 'table.header.complianceType',
        resourceKey: 'complianceType',
        key: 'complianceType',
      },
      {
        msgKey: 'description.title.api.version',
        resourceKey: 'apiVersion',
        key: 'apiVersion',
      },
      {
        msgKey: 'description.title.last.transition',
        resourceKey: 'lastTransition',
        key: 'lastTransition',
      },
      {
        msgKey: 'table.header.compliant',
        resourceKey: 'compliant',
        key: 'compliant',
        transformFunction: getStatus
      },
    ],
  },
  policyRoleBindingTemplates: {
    title: 'table.header.roleBindingTemplates',
    defaultSortField: 'name',
    normalizedKey: 'name',
    resourceKey: 'roleBindingTemplates',
    tableKeys: [
      {
        msgKey: 'table.header.name',
        resourceKey: 'name',
        key: 'name',
      },
      {
        msgKey: 'table.header.complianceType',
        resourceKey: 'complianceType',
        key: 'complianceType',
      },
      {
        msgKey: 'description.title.api.version',
        resourceKey: 'apiVersion',
        key: 'apiVersion',
      },
      {
        msgKey: 'description.title.last.transition',
        resourceKey: 'lastTransition',
        key: 'lastTransition',
      },
      {
        msgKey: 'table.header.compliant',
        resourceKey: 'compliant',
        key: 'compliant',
        transformFunction: getStatus
      },
    ],
  },
  policyObjectTemplates: {
    title: 'table.header.objectTemplates',
    defaultSortField: 'name',
    normalizedKey: 'name',
    resourceKey: 'objectTemplates',
    tableKeys: [
      {
        msgKey: 'table.header.name',
        resourceKey: 'name',
        key: 'name',
      },
      {
        msgKey: 'table.header.complianceType',
        resourceKey: 'complianceType',
        key: 'complianceType',
      },
      {
        msgKey: 'description.title.api.version',
        resourceKey: 'apiVersion',
        key: 'apiVersion',
      },
      {
        msgKey: 'description.title.last.transition',
        resourceKey: 'lastTransition',
        key: 'lastTransition',
      },
      {
        msgKey: 'table.header.compliant',
        resourceKey: 'compliant',
        key: 'compliant',
        transformFunction: getStatus
      },
    ],
  },
  policyViolations: {
    resourceKey: 'violations',
    title: 'table.header.violation',
    defaultSortField: 'name',
    normalizedKey: 'name',
    tableKeys: [
      {
        msgKey: 'table.header.status',
        resourceKey: 'status',
        key: 'status',
        transformFunction: getStatus,
      },
      {
        msgKey: 'table.header.cluster',
        resourceKey: 'cluster',
        key: 'cluster',
      },
      {
        msgKey: 'table.header.name',
        resourceKey: 'name',
        key: 'name',
      },
      {
        msgKey: 'table.header.message',
        resourceKey: 'message',
        key: 'message',
      },
      {
        msgKey: 'table.header.reason',
        resourceKey: 'reason',
        key: 'reason',
      },
      {
        msgKey: 'table.header.selector',
        resourceKey: 'selector',
        key: 'selector',
        transformFunction: getLabelsToList
      },
    ],
  },
  tableKeys: [
    {
      msgKey: 'table.header.name',
      resourceKey: 'metadata.name',
      transformFunction: createPolicyLink,
    },
    {
      msgKey: 'table.header.namespace',
      resourceKey: 'metadata.namespace',
    },
    {
      msgKey: 'table.header.status',
      resourceKey: 'status',
      transformFunction: getStatus,
    },
    {
      msgKey: 'table.header.enforcement',
      resourceKey: 'enforcement',
      transformFunction: getEnforcement,
    },
  ],
  tableActions: [
    'table.actions.remove',
  ],
  detailKeys: {
    title: 'policy.details',
    headerRows: ['type', 'detail'],
    rows: [
      {
        cells: [
          {
            resourceKey: 'description.title.name',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.name'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.namespace',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.namespace'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.status',
            type: 'i18n'
          },
          {
            resourceKey: 'status'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'table.header.message',
            type: 'i18n'
          },
          {
            resourceKey: 'message'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.exclude_namespace',
            type: 'i18n'
          },
          {
            resourceKey: 'detail.exclude_namespace',
            transformFunction: getExcludeNamespace
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.include_namespace',
            type: 'i18n'
          },
          {
            resourceKey: 'detail.include_namespace',
            transformFunction: getIncludeNamespace
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.enforcement',
            type: 'i18n'
          },
          {
            resourceKey: 'enforcement'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.created',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.creationTimestamp',
            transformFunction: getAge
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.annotations',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.annotations',
            transformFunction: getLabelsToList
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.resource.version',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.resourceVersion'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.self.link',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.selfLink'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.uid',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.uid'
          }
        ]
      },
    ]
  },
}

export function createPolicyLink(item = {}, ...param){
  if (param[2]) return item.metadata.name
  return (item && item.metadata) ? <Link to={`${config.contextPath}/local/${encodeURIComponent(item.metadata.namespace)}/${encodeURIComponent(item.metadata.name)}`}>{item.metadata.name}</Link> : <Link to={`${config.contextPath}/local/`}>{JSON.stringify(item)}</Link>
}

export function getStatus(item= {}, locale) {
  const expectedStatuses = [ 'compliant', 'notcompliant', 'noncompliant', 'invalid']
  if (item) {
    if (item.status && expectedStatuses.indexOf(item.status.toLowerCase()) > -1){
      if (item.status.toLowerCase() === 'compliant') {
        return <StatusField status='ok' text={msgs.get('policy.status.compliant', locale)} />
      } else {
        return <StatusField status='critical' text={msgs.get(`policy.status.${item.status.toLowerCase()}`, locale)} />
      }
    }
    if (item.compliant && expectedStatuses.indexOf(item.compliant.toLowerCase()) > -1){
      if (item.compliant.toLowerCase() === 'compliant') {
        return <StatusField status='ok' text={msgs.get('policy.status.compliant', locale)} />
      } else {
        return <StatusField status='critical' text={msgs.get(`policy.status.${item.status.toLowerCase()}`, locale)} />
      }
    }
  }
  return '-'
}

export function getExcludeNamespace(item) {
  const namespace = lodash.get(item, 'detail.exclude_namespace')
  if (namespace) {
    return namespace.join(', ')
  }
  return '-'
}

export function getIncludeNamespace(item) {
  const namespace = lodash.get(item, 'detail.include_namespace')
  if (namespace) {
    return namespace.join(', ')
  }
  return '-'
}

export function getEnforcement(item, locale) {
  const expectedEnforcements = [ 'enforce', 'inform']
  if (item.enforcement&&expectedEnforcements.indexOf(item.enforcement.toLowerCase()) > -1){
    return msgs.get(`policy.enforcement.${item.enforcement.toLowerCase()}`, locale)
  }
  return '-'
}

export function getAPIGroups(item) {
  const apiGroups = lodash.get(item, 'apiGroups')
  if (apiGroups) {
    return apiGroups.join(', ')
  }
  return '-'
}

export function getRuleVerbs(item) {
  const verbs = lodash.get(item, 'verbs')
  if (verbs) {
    return verbs.join(', ')
  }
  return '-'
}
