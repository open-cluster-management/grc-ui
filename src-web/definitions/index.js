/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import lodash from 'lodash'

import { RESOURCE_TYPES } from '../../lib/shared/constants'

import hcmpoliciescluster from './hcm-policies-cluster'
import hcmpoliciespolicy from './hcm-policies-policy'
import hcmcompliances from './hcm-compliances'
import hcmpolicies from './hcm-policies'
import hcmsecurityfindings from './hcm-security-findings'
import hcmclusterfindings from './hcm-cluster-findings'
import hcmpoliciesapplication from './hcm-policies-application'

const resourceData = {
  [RESOURCE_TYPES.HCM_POLICIES_PER_APPLICATION.name]: hcmpoliciesapplication,
  [RESOURCE_TYPES.HCM_POLICIES_PER_POLICY.name]: hcmpoliciespolicy,
  [RESOURCE_TYPES.HCM_POLICIES_PER_CLUSTER.name]: hcmpoliciescluster,
  [RESOURCE_TYPES.HCM_COMPLIANCES.name]: hcmcompliances,
  [RESOURCE_TYPES.HCM_POLICIES.name]: hcmpolicies,
  [RESOURCE_TYPES.HCM_SECURITY_FINDINGS.name]: hcmsecurityfindings,
  [RESOURCE_TYPES.HCM_CLUSTER_FINDINGS.name]: hcmclusterfindings,
}

// diagram tabs
function getResourceData(resourceType) {
  // main Topology tab
  const def = resourceData[resourceType.name]
  if (!def) {
    //eslint-disable-next-line no-console
    console.error(`No resource data found for '${JSON.stringify(resourceType)}'`)
  }
  return def
}

export default getResourceData

export function getPrimaryKey(resourceType) {
  let pk = 'metadata.uid'
  const def = getResourceData(resourceType)
  if (def && def.primaryKey) {
    pk = def.primaryKey
  }
  return pk
}

export function getSecondaryKey(resourceType) {
  let sk = 'cluster'
  const def = getResourceData(resourceType)
  if(def && def.secondaryKey) {
    sk = def.secondaryKey
  }
  return sk
}

export function getURIKey(resourceType) {
  let uriKey = 'metadata.name'
  const def = getResourceData(resourceType)
  if (def && def.uriKey) {
    uriKey = def.uriKey
  }
  return uriKey
}

export function getDefaultSearchField(resourceType) {
  const def = getResourceData(resourceType)
  let sf = def && def.defaultSearchField
  if (!def || !def.tableKeys || (def && def.tableKeys.length < 1))
    //eslint-disable-next-line no-console
    console.error(`No table keys found in ${resourceType} resource definition`)
  if (!sf)
    sf = def && def.tableKeys && def.tableKeys[0] && def.tableKeys[0].resourceKey
  return sf
}

export function getDefaultSortField(resourceType) {
  const def = getResourceData(resourceType)
  let sf = def && def.defaultSortField
  if (!def || !def.tableKeys || (def && def.tableKeys.length < 1))
    //eslint-disable-next-line no-console
    console.error(`No table keys found in ${resourceType} resource definition`)
  if (!sf) {
    sf = def && def.tableKeys && def.tableKeys[0] && def.tableKeys[0].resourceKey
  }
  if (!sf)
    //eslint-disable-next-line no-console
    console.error(`No sortable fields defined for '${resourceType}' resource definition`)
  return sf
}

export function getTableKeys(resourceType) {
  const def = getResourceData(resourceType)
  return def.tableKeys
}

export function getLink(link, resource) {
  if(typeof(link) === 'boolean') {
    return `/${resource.metadata.namespace}/${resource.metadata.name}`
  } else if (typeof(link) === 'function') {
    return link(resource)
  } else {
    const parts = link.split('/')
    let path = ''
    parts.forEach(part => {
      const value = lodash.get(resource, part)
      path += `/${encodeURIComponent(value)}`
    })
    return path
  }
}
