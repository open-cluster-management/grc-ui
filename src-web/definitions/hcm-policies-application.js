/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */

'use strict'

import React from 'react'
import config from '../../lib/shared/config'
import { getTruncatedText } from './hcm-policies-cluster'

export default {
  defaultSortField: 'name',
  primaryKey: 'name',
  secondaryKey: 'nameSpace',
  tableActions: [
    'table.actions.policy.applications.sidepanel',
  ],
  tableKeys: [
    {
      msgKey: 'table.header.application',
      resourceKey: 'name',
      transformFunction: createApplicationLink,
    },
    {
      msgKey: 'string.namespace',
      resourceKey: 'nameSpace',
    },
    {
      msgKey: 'table.header.violation',
      resourceKey: 'violations',
    },
    {
      msgKey: 'table.header.violated',
      resourceKey: 'violatedPolicies',
      transformFunction: getAppViolationLabels,
    }
  ],
  applicationViolatedSidePanel: {
    headerRows: ['', 'table.header.policy.name', 'table.header.rule.violation', 'table.header.control', 'table.header.cluster',],
    subHeaders: ['string.name', 'table.header.message', 'table.header.reason', '',],
    rows: [
      {
        cells: [
          {
            resourceKey: 'metadata.name',
          },
          {
            resourceKey: 'violatedNum',
          },
          {
            resourceKey: 'metadata.annotations["policy.open-cluster-management.io/controls"]',
          },
          {
            resourceKey: 'cluster',
          },
        ]
      }
    ]
  },
}

export function getAppViolationLabels(item) {
  if (item && Array.isArray(item.violatedPolicies) && item.violatedPolicies.length > 0){
    let violatedPoliciesString = ''
    item.violatedPolicies.forEach((violatedPolicy, index) => {
      if(violatedPolicy.name && typeof violatedPolicy.name === 'string') {
        if (index+1 < item.violatedPolicies.length) {
          violatedPoliciesString = `${violatedPoliciesString}${violatedPolicy.name}, `
        }
        else {
          violatedPoliciesString = `${violatedPoliciesString}${violatedPolicy.name}`
        }
      }
    })
    return getTruncatedText(violatedPoliciesString)
  }
  return '-'
}

export function createApplicationLink(item){
  if (item && item.name && item.nameSpace) {
    return <a href={`${config.contextPath}/applications/${item.nameSpace}/${item.name}`}>{item.name}</a>
  }
  else if (item && item.name) {
    return item.name
  }
  return '-'
}
