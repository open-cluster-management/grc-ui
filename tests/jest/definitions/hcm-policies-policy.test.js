/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'
import { createComplianceLink } from '../../../src-web/definitions/hcm-policies-policy'
import { hcmPoliciesPolicyItem } from './DefinitionsTestingData'

describe('createComplianceLink', () => {
  it('Should item.metadata.name', () => {
    expect(createComplianceLink(hcmPoliciesPolicyItem, 'en-US', 'test', 'grcTest')).toMatchSnapshot()
  })
  it('Should create policy link', () => {
    expect(createComplianceLink(hcmPoliciesPolicyItem, 'en-US')).toMatchSnapshot()
  })
  it('Should create compliance link', () => {
    hcmPoliciesPolicyItem.raw.kind = 'Compliance'
    expect(createComplianceLink(hcmPoliciesPolicyItem, 'en-US')).toMatchSnapshot()
  })
  it('Should undefined', () => {
    expect(createComplianceLink()).toMatchSnapshot()
  })
})
