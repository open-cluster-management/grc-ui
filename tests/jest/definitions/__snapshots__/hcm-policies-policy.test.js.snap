// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createComplianceLink Should create compliance link 1`] = `
<Link
  replace={false}
  to="undefined/all/policy1"
>
  policy1
   (Deprecated)
</Link>
`;

exports[`createComplianceLink Should create policy link 1`] = `
<Link
  replace={false}
  to="undefined/all/mcm/policy1"
>
  policy1
</Link>
`;

exports[`createComplianceLink Should item.metadata.name 1`] = `"policy1"`;

exports[`createComplianceLink Should undefined 1`] = `undefined`;
